src/dstore/pmix_esh.c

```diff
@@ -144,23 +175,251 @@ static inline const char *_unique_id(void)
     return str;
 }
 
+static inline int _esh_nspace_tbl_init(void)
+{
+    _esh_nspace_cnt = ESH_INIT_NSPACE_CNT;
+    _esh_nspace_tbl_size = ESH_INIT_NSPACE_TBL_SIZE;
+
+    _nspace_tbl = (nspace_tbl_entry_t*)malloc(sizeof(nspace_tbl_entry_t) * _esh_nspace_tbl_size);
+    if (NULL == _nspace_tbl) {
+        PMIX_ERROR_LOG(PMIX_ERROR);
+        return PMIX_ERROR;
+    }
+    memset(_nspace_tbl, 0x00, sizeof(nspace_tbl_entry_t) * _esh_nspace_tbl_size);
+
+    _nspace_map = (nspace_entry_t*)malloc(sizeof(nspace_entry_t) * _esh_nspace_cnt);
+    if (NULL == _nspace_map) {
+        PMIX_ERROR_LOG(PMIX_ERROR);
+        return PMIX_ERROR;
+    }
+    memset(_nspace_map, 0x00, sizeof(nspace_entry_t) * _esh_nspace_cnt);
+
+    return PMIX_SUCCESS;
+}
+
+static inline void _esh_nspace_tbl_finalize(void)
+{
+    int idx;
+
+    if (NULL != _nspace_tbl) {
+        for (idx = 0; idx < _esh_nspace_tbl_size; idx++) {
+            if (!_nspace_tbl[idx].in_use) {
+                continue;
+            }
+            _nspace_tbl[idx].in_use = 0;
+            close(_nspace_tbl[idx].lockfd);
+
+            if (NULL != _nspace_tbl[idx].lockfile) {
+                if(_is_server()) {
+                    unlink(_nspace_tbl[idx].lockfile);
+                }
+                free(_nspace_tbl[idx].lockfile);
+                _nspace_tbl[idx].lockfile = NULL;
+            }
+            if (NULL != _nspace_tbl[idx].nspace_path) {
+                if(_is_server()) {
+                    rmdir(_nspace_tbl[idx].nspace_path);
+                }
+                free(_nspace_tbl[idx].nspace_path);
+                _nspace_tbl[idx].nspace_path = NULL;
+            }
+        }
+        _esh_nspace_tbl_size = 0;
+        free(_nspace_tbl);
+    }
+
+    if (NULL != _nspace_map) {
+        for (idx = 0; idx < _esh_nspace_cnt; idx++) {
+            if (NULL == _nspace_map[idx].nspace) {
+                continue;
+            }
+            free(_nspace_map[idx].nspace);
+            _nspace_map[idx].nspace = NULL;
+        }
+        _esh_nspace_cnt = 0;
+        free(_nspace_map);
+    }
+}
+
+static inline int _esh_nspace_tbl_map_idx(const char *nspace, int tbl_idx)
+{
+    int i;
+    for (i = 0; i < _esh_nspace_cnt; i++) {
+        if (NULL == _nspace_map[i].nspace) {
+            _nspace_map[i].nspace = strdup(nspace);
+            _nspace_map[i].tbl_idx = tbl_idx;
+            return PMIX_SUCCESS;
+        }
+    }
+    _nspace_map = (nspace_entry_t *)realloc(_nspace_map,
+                        sizeof(nspace_entry_t) * _esh_nspace_cnt * 2);
+    if (NULL == _nspace_map) {
+        PMIX_ERROR_LOG(PMIX_ERROR);
```

Comment:
Use more detailed `PMIX_ERR_OUT_OF_RESOURCE` errcode instead of universal `PMIX_ERROR`